src/ocf_core.c

```diff
@@ -221,137 +221,122 @@ static void ocf_req_complete(struct ocf_request *req, int error)
+		req->cache_mode = ocf_req_cache_mode_fast;
+	}
 
-	ocf_io_get(io);
+	if (cache->trace.trace_callback) {
+		if (io->dir == OCF_WRITE)
+			ocf_trace_prep_io_event(&trace_event, req, ocf_event_operation_wr);
+		else if (io->dir == OCF_READ)
+			ocf_trace_prep_io_event(&trace_event, req, ocf_event_operation_rd);
+	}
 
-	ret = ocf_engine_hndl_req(req);
-	if (ret) {
-		dec_counter_if_req_was_dirty(req);
-		ocf_io_end(io, ret);
+	fast = ocf_engine_hndl_fast_req(req);
+	if (fast != OCF_FAST_PATH_NO) {
+		ocf_trace_push(io->io_queue, &trace_event, sizeof(trace_event));
+		return 0;
 	}
+
+	ret = -OCF_ERR_IO;
+map_allocated:
+	ocf_req_dealloc_map(req);
+error:
+	dec_counter_if_req_was_dirty(req);
+	return ret;
 }
 
-int ocf_core_submit_io_fast(struct ocf_io *io)
+void ocf_core_volume_submit_io(struct ocf_io *io)
 {
 	struct ocf_request *req;
-	struct ocf_event_io trace_event;
 	ocf_core_t core;
 	ocf_cache_t cache;
-	int fast;
 	int ret;
 
 	OCF_CHECK_NULL(io);
 
 	ret = ocf_core_validate_io(io);
-	if (ret < 0)
-		return ret;
+	if (ret < 0) {
+		ocf_io_end(io, ret);
+		return;
+	}
 
 	req = ocf_io_to_req(io);
 	core = ocf_volume_to_core(ocf_io_get_volume(io));
 	cache = ocf_core_get_cache(core);
 
+	ocf_trace_init_io(req);
+
 	if (unlikely(!env_bit_test(ocf_cache_state_running,
-			&cache->cache_state))) {
+					&cache->cache_state))) {
 		ocf_io_end(io, -OCF_ERR_CACHE_NOT_AVAIL);
-		return 0;
-	}
-
-	if (req->d2c) {
-		dec_counter_if_req_was_dirty(req);
-		return -OCF_ERR_IO;
-	}
-
-	ret = ocf_req_alloc_map(req);
-	if (ret) {
-		ocf_io_end(io, -OCF_ERR_NO_MEM);
-		return 0;
+		return;
 	}
 
+	req->part_id = ocf_user_part_class2id(cache, io->io_class);
 	req->core = core;
 	req->complete = ocf_req_complete;
-	req->part_id = ocf_user_part_class2id(cache, io->io_class);
 
 	ocf_resolve_effective_cache_mode(cache, core, req);
 
-	switch (req->cache_mode) {
-	case ocf_req_cache_mode_pt:
-		return -OCF_ERR_IO;
-	case ocf_req_cache_mode_wb:
-	case ocf_req_cache_mode_wo:
-		req->cache_mode = ocf_req_cache_mode_fast;
-		break;
-	default:
-		if (cache->use_submit_io_fast)
-			break;
-		if (io->dir == OCF_WRITE)
-			return -OCF_ERR_IO;
-
-		req->cache_mode = ocf_req_cache_mode_fast;
-	}
-
 	ocf_core_update_stats(core, io);
 
-	if (cache->trace.trace_callback) {
-		if (io->dir == OCF_WRITE)
-			ocf_trace_prep_io_event(&trace_event, req, ocf_event_operation_wr);
-		else if (io->dir == OCF_READ)
-			ocf_trace_prep_io_event(&trace_event, req, ocf_event_operation_rd);
-	}
-
 	ocf_io_get(io);
 
-	fast = ocf_engine_hndl_fast_req(req);
-	if (fast != OCF_FAST_PATH_NO) {
-		ocf_trace_push(io->io_queue, &trace_event, sizeof(trace_event));
+	if (!ocf_core_submit_io_fast(io, req, core, cache)) {
 		ocf_core_seq_cutoff_update(core, req);
-		return 0;
+		return;
 	}
 
-	dec_counter_if_req_was_dirty(req);
+	// Recalculate cache mode if fast path failed and switched it to ocf_req_cache_mode_fast
+	ocf_resolve_effective_cache_mode(cache, core, req);
```

Comment:
This is rather expensive, because it performs seq_cutoff check. Maybe fast path could restore original cache mode on fail?